qiskit/providers/aer/backends/aerbackend.py

```diff
@@ -366,6 +371,10 @@ def _assemble(self, circuits, parameter_binds=None, **run_options):
             else:
                 qobj = assemble(circuits, backend=self)
 
+            # Recover metadata of circuits
+            for circuit, metadata in zip(circuits, metadata_list):
+                circuit.metadata = metadata
```

Comment:
So I think the issue with doing this is that the metadata doesn't get set in the output result. The idea behind metadata is that it provide more rich data around tracking the purpose of a circuit in a larger application. Prior to the introduction of `circuit.metadata` the only available information was the `circuit.name` which let to people trying to encode this all in a string. 

While metadata shouldn't influence the execution of the circuit on a backend it is important that users can correlate the results to the metadata easily (similar to how you can find a circuit name associated with a result object). Is there a way we can preserve setting the metadata in the result like it was before, but skip the json serialization in `assemble()`?

Reply:
I think there are two ways:

Approach-1: take metadata from `QuantumCircuit` before `assemble()` and store the metadata to a result before `AerJob.result()` in a caller thread
Approach-2: take metadata from `QObj` before `_execute()` and store the metadata to a result after `_execute()` in pooled threads

In approach-1, the taken metadata needs to be kept somewhere. I believe Approach-2 looks better and implemented it in fd65ac7.